File: app/src/main/java/com/riders/thelab/ui/floatinglabels/FloatingLabelsActivity.java
package com.riders.thelab.ui.floatinglabels;

import android.annotation.SuppressLint;
import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.View;
import android.view.WindowManager;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.Toast;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputLayout;
import com.riders.thelab.R;
import com.riders.thelab.core.utils.LabCompatibilityManager;
import com.riders.thelab.ui.base.SimpleActivity;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;


/**
 * Created by Michaël on 07/03/2017.
 */

@SuppressLint("NonConstantResourceId")
public class FloatingLabelsActivity extends SimpleActivity {

    //private Toolbar mFloatingLabelsToolbar;
    @BindView(R.id.input_name)
    EditText inputName;
    @BindView(R.id.input_email)
    EditText inputEmail;
    @BindView(R.id.input_password)
    EditText inputPassword;
    @BindView(R.id.input_layout_name)
    TextInputLayout inputLayoutName;
    @BindView(R.id.input_layout_email)
    TextInputLayout inputLayoutEmail;
    @BindView(R.id.input_layout_password)
    TextInputLayout inputLayoutPassword;
    @BindView(R.id.floating_labels_btn_signup)
    MaterialButton btnSignUp;

    private static boolean isValidEmail(String email) {
        return !TextUtils.isEmpty(email)
                && android.util.Patterns.EMAIL_ADDRESS.matcher(email).matches();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (LabCompatibilityManager.isTablet(this))
            //Force screen orientation to Landscape mode
            this.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
        else
            //Force screen orientation to Portrait mode
            this.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);

        setContentView(R.layout.activity_floating_labels);

        ButterKnife.bind(this);

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setTitle(getString(R.string.activity_title_floating_labels));

        inputName.addTextChangedListener(new MyTextWatcher(inputName));
        inputEmail.addTextChangedListener(new MyTextWatcher(inputEmail));
        inputPassword.addTextChangedListener(new MyTextWatcher(inputPassword));
    }

    @OnClick(R.id.floating_labels_btn_signup)
    void onButtonClicked() {
        try {
            InputMethodManager imm = (InputMethodManager) getSystemService(INPUT_METHOD_SERVICE);
            if (imm != null) {
                imm.hideSoftInputFromWindow(getCurrentFocus().getWindowToken(), 0);
            }
        } catch (Exception e) {
            e.getMessage();
        }

        submitForm();

    }

    /**
     * Validating form
     */
    private void submitForm() {
        if (!validateName()) {
            return;
        }

        if (!validateEmail()) {
            return;
        }

        if (!validatePassword()) {
            return;
        }

        Toast.makeText(getApplicationContext(), "Thank You!", Toast.LENGTH_SHORT).show();
    }

    private boolean validateName() {
        if (inputName.getText().toString().trim().isEmpty()) {
            inputLayoutName.setError(getString(R.string.err_msg_name));
            requestFocus(inputName);
            return false;
        } else {
            inputLayoutName.setErrorEnabled(false);
        }

        return true;
    }

    private boolean validateEmail() {
        String email = inputEmail.getText().toString().trim();

        if (email.isEmpty() || !isValidEmail(email)) {
            inputLayoutEmail.setError(getString(R.string.err_msg_email));
            requestFocus(inputEmail);
            return false;
        } else {
            inputLayoutEmail.setErrorEnabled(false);
        }

        return true;
    }

    private boolean validatePassword() {
        if (inputPassword.getText().toString().trim().isEmpty()) {
            inputLayoutPassword.setError(getString(R.string.err_msg_password));
            requestFocus(inputPassword);
            return false;
        } else {
            inputLayoutPassword.setErrorEnabled(false);
        }

        return true;
    }

    private void requestFocus(View view) {
        if (view.requestFocus()) {
            getWindow()
                    .setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
        }
    }


    private class MyTextWatcher implements TextWatcher {

        private final View view;

        private MyTextWatcher(View view) {
            this.view = view;
        }

        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {

        }

        @Override
        public void afterTextChanged(Editable sEditable) {
            switch (view.getId()) {
                case R.id.input_name:
                    validateName();
                    break;

                case R.id.input_email:
                    validateEmail();
                    break;

                case R.id.input_password:
                    validatePassword();
                    break;

                default:
                    break;
            }
        }
    }

}

File: app/src/main/java/com/riders/thelab/ui/floatinglabels/FloatingLabelsActivity.kt
package com.riders.thelab.ui.floatinglabels

import android.annotation.SuppressLint
import android.content.pm.ActivityInfo
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.WindowManager
import android.view.inputmethod.InputMethodManager
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.riders.thelab.R
import com.riders.thelab.core.utils.LabCompatibilityManager
import com.riders.thelab.databinding.ActivityFloatingLabelsBinding
import com.riders.thelab.utils.Validator
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
open class FloatingLabelsActivity : AppCompatActivity() {

    lateinit var viewBinding: ActivityFloatingLabelsBinding

    @SuppressLint("SourceLockedOrientationActivity")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        //Force screen orientation to Landscape mode
        if (LabCompatibilityManager.isTablet(this))
            this.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
        //Force screen orientation to Portrait mode
        else
            this.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT

        viewBinding = ActivityFloatingLabelsBinding.inflate(layoutInflater)
        setContentView(viewBinding.root)

        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        supportActionBar!!.setTitle(getString(R.string.activity_title_floating_labels))

        viewBinding.inputName.addTextChangedListener(MyTextWatcher(viewBinding.inputName))
        viewBinding.inputEmail.addTextChangedListener(MyTextWatcher(viewBinding.inputEmail))
        viewBinding.inputPassword.addTextChangedListener(MyTextWatcher(viewBinding.inputPassword))

        viewBinding.floatingLabelsBtnSignup.setOnClickListener {
            try {
                val imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
                imm.hideSoftInputFromWindow(currentFocus!!.windowToken, 0)
            } catch (e: Exception) {
                e.message
            }
            submitForm()
        }
    }

    /**
     * Validating form
     */
    private fun submitForm() {
        if (!validateName()) {
            return
        }
        if (!validateEmail()) {
            return
        }
        if (!validatePassword()) {
            return
        }
        Toast.makeText(applicationContext, "Thank You!", Toast.LENGTH_SHORT).show()
    }

    open fun validateName(): Boolean {
        if (viewBinding.inputName.text.toString().trim { it <= ' ' }.isEmpty()) {
            viewBinding.inputLayoutName.error = getString(R.string.err_msg_name)
            requestFocus(viewBinding.inputName)
            return false
        } else {
            viewBinding.inputLayoutName.isErrorEnabled = false
        }
        return true
    }

    open fun validateEmail(): Boolean {
        val email: String = viewBinding.inputEmail.text.toString().trim { it <= ' ' }
        if (email.isEmpty() || !Validator.isValidEmail(email)) {
            viewBinding.inputLayoutEmail.error = getString(R.string.err_msg_email)
            requestFocus(viewBinding.inputEmail)
            return false
        } else {
            viewBinding.inputLayoutEmail.isErrorEnabled = false
        }
        return true
    }

    open fun validatePassword(): Boolean {
        if (viewBinding.inputPassword.text.toString().trim { it <= ' ' }.isEmpty()) {
            viewBinding.inputLayoutPassword.error = getString(R.string.err_msg_password)
            requestFocus(viewBinding.inputPassword)
            return false
        } else {
            viewBinding.inputLayoutPassword.isErrorEnabled = false
        }
        return true
    }

    private fun requestFocus(view: View) {
        if (view.requestFocus()) {
            window
                .setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE)
        }
    }


    internal class MyTextWatcher(private val view: View) : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
        override fun afterTextChanged(sEditable: Editable) {
            when (view.id) {
                R.id.input_name -> FloatingLabelsActivity().validateName()
                R.id.input_email -> FloatingLabelsActivity().validateEmail()
                R.id.input_password -> FloatingLabelsActivity().validatePassword()
            }
        }
    }
}